File: 100/143 - Reorder List.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
void reorderList(struct ListNode* head) {
    struct ListNode *temp = head, *last = head, *curr;
    
    if (! (head && head->next))
        return;
    
    while (1) {
        if (last->next)
            last = last->next;
        else break;
        
        if (last->next)
            last = last->next;
        else break;
        
        temp = temp->next;
    }
    
    last = temp->next;
    temp->next = NULL;
    curr = last->next;
    last->next = NULL;
    while (curr) {
        temp = curr->next;
        curr->next = last;
        last = curr;
        curr = temp;
    }
    
    curr = head;
    while (last) {
        temp = curr->next;
        curr->next = last;
        curr = temp;
        temp = last->next;
        last->next = curr;
        last = temp;
    }
}

File: csharp/143.reorder-list.cs
/*
 * @lc app=leetcode id=143 lang=csharp
 *
 * [143] Reorder List
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void ReorderList(ListNode head) {
        ListNode slow = head, fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        if (slow == null) {
            return;
        }
        ListNode curr = slow.next;
        slow.next = null;
        fast = null;
        while (curr != null) {
            ListNode tmp = curr.next;
            curr.next = fast;
            fast = curr;
            curr = tmp;
        }
        while (fast != null) {
            ListNode tmp = fast.next;
            slow = head.next;
            head.next = fast;
            fast.next = slow;
            head = slow;
            fast = tmp;
        }
    }
}

